api/types/plugin.go

```diff
@@ -0,0 +1,254 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package types
+
+import (
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/gravitational/trace"
+)
+
+// Plugin represents a plugin instance
+type Plugin interface {
+	// ResourceWithSecrets provides common resource methods.
+	ResourceWithSecrets
+	Clone() Plugin
+	GetCredentials() PluginCredentials
+	SetCredentials(PluginCredentials) error
+	SetStatus(PluginStatus) error
+}
+
+// PluginCredentials are the credentials embedded in Plugin
+type PluginCredentials interface {
+	GetOauth2AccessToken() *PluginOAuth2AccessTokenCredentials
+}
+
+// PluginStatus is the plugin status
+type PluginStatus interface {
+	GetCode() PluginStatusCode
+}
+
+// NewPluginV1 creates a new PluginV1 resource.
+func NewPluginV1(name string, spec PluginSpecV1, creds *PluginCredentialsV1) *PluginV1 {
+	p := &PluginV1{
+		Metadata: Metadata{
+			Name: name,
+		},
+		Spec: spec,
+	}
+	if creds != nil {
+		p.SetCredentials(creds)
+	}
+
+	return p
+}
+
+// CheckAndSetDefaults checks validity of all parameters and sets defaults.
+func (p *PluginV1) CheckAndSetDefaults() error {
+	p.setStaticFields()
+
+	if err := p.Metadata.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if p.Kind == "" {
+		return trace.BadParameter("resource has an empty Kind field")
+	}
+	if p.Version == "" {
+		return trace.BadParameter("resource has an empty Version field")
+	}
+	if p.Spec.Settings == nil {
+		return trace.BadParameter("settings must be set")
+	}
```

Comment:
Should be moved to a `default:` case in the switch right below, IMO.

The message can be "unknown or missing settings".